### PR TITLE
Convert canvas thumbnails to PNG

### LINKED ISSUE #6035: Reduce memory footprint of thumbnails
We have discussed on IRC that it might be helpful to package rendered thumbnails as PNG to reduce their memory consumption. This should be done when the viewer is idle, i.e., not when the user is scrolling.

### PR DESCRIPTION
**Pull request by zinking**
fixes #6035 , converting canvas thumb nails to png images

### CODE REVIEW

File: web/pdf_thumbnail_view.js [hunk 1]
```diff
@@ -180,15 +180,15 @@ var PDFThumbnailView = (function PDFThumbnailViewClosure() {
...
-      canvas.setAttribute('aria-label', mozL10n.get('thumb_page_canvas',
-        {page: this.id}, 'Thumbnail of Page {{page}}'));
+      //canvas.className = 'thumbnailImage';
+      //canvas.setAttribute('aria-label', mozL10n.get('thumb_page_canvas',
+      //  {page: this.id}, 'Thumbnail of Page {{page}}'));
 
       this.canvas = canvas;
       this.div.setAttribute('data-loaded', true);
-      this.ring.appendChild(canvas);
+      //this.ring.appendChild(canvas);
```
Comment (timvandermeij) [inline web/pdf_thumbnail_view.js]: Please remove the commented-out code here and above.

File: .gitignore [hunk 2]
```diff
@@ -6,3 +6,4 @@ tags
 Makefile
 node_modules/
 examples/node/svgdump/
+.idea
```
Comment (timvandermeij) [inline .gitignore]: Please revert this change as it is not related to the patch.

File: web/pdf_thumbnail_view.js [hunk 3]
```diff
@@ -199,6 +199,14 @@ var PDFThumbnailView = (function PDFThumbnailViewClosure() {
       if (!noCtxScale && outputScale.scaled) {
         ctx.scale(outputScale.sx, outputScale.sy);
       }
+      this.image  = document.createElement('img');
+      this.image.id = this.renderingId;
+      this.image.style.height = canvas.style.height;
+      this.image.style.width = canvas.style.width;
+      this.image.className = 'thumbnailImage';
+      this.image.setAttribute('aria-label', mozL10n.get('thumb_page_canvas',
+            {page: this.id}, 'Thumbnail of Page {{page}}'));
```
Comment (timvandermeij) [inline web/pdf_thumbnail_view.js]: Indent this line with only two spaces instead of six. Also add spaces before `}` and after `{` to get `{ page: this.id }`.

**Comment by timvandermeij** [comment]
@zinking Once you have addressed the review comment, please squash the commits into one commit. See https://github.com/mozilla/pdf.js/wiki/Squashing-Commits on how to do that easily.

**Comment by Snuffleupagus** [comment]
It appears that the following from https://github.com/mozilla/pdf.js/issues/6035#issue-78226866 isn't implemented by the PR yet:

> This should be done when the viewer is idle, i.e., not when the user is scrolling.

**Comment by yurydelendik** [comment]
The solution looks fine to me. @Snuffleupagus I think we can leave without reencoding-on-cleanup part -- unless there is a clear indication the encoding of the png takes unreasonable amount of CPU. It will use some CPU but I'm thinking in overall not using lots of canvases might be a better choice. It will be better if we will have some benchmark with this PR.

Also, using canvas'es toBlob() instead of toDataURL in some way might help with memory usage (similar to PDFJS.createObjectURL but it shall be async).

**Comment by timvandermeij** [comment]
Looks good to me too with the nit addressed and re-squashed after that. `toBlob` might be better indeed, but having a benchmark to confirm that would be nice. Does not have to be too exact, just measuring the process memory usage a couple of times and taking the average before and after the patch should be enough.

**Comment by timvandermeij** [comment]
@zinking Do you have time to finish up this PR by addressing the above comments?

**Comment by zinking** [comment]
@timvandermeij  do you mean manually testing the memory usage ?   I can do the rough testing using a few documents at hand, and provide some statistics, but this `comment` probably should be framed better so that my manual test is useful

**Comment by timvandermeij** [comment]
Yes, all that needs to be done is address the nit above and manually test the memory usage before and after your patch.

**Comment by zinking** [comment]
browser	|pdf	|Page	|before(private)	|After (private)	|comment
-------------|------|-------------|--------------------------|-------------------|-------------
chrome	|compressed.tracemonkey-pldi-09.pdf	|14	|406,260k	|158,820k	|
chrome	|abel.pdf	|285	|638,996k	|394,564k|	this has more pics
chrome	|AkkaScala.pdf	|387	|768,476k	|434,732k	|

The test is conducted on OSX 10.9.5 , Chrome 44.0.2403.157 (64 bit) 
memory stats is generated from : chrome://memory-redirect/

looks saving memory as expected.

**Comment by timvandermeij** [comment]
@zinking Great, thank you for confirming that! Could you squash the commits into one commit? Then this should be good to merge.

**Comment by timvandermeij** [comment]
Thank you for the patch!
